builtin/logical/database/backend.go

```diff
@@ -228,6 +272,156 @@ func (b *databaseBackend) GetConnection(ctx context.Context, s logical.Storage,
+// so initializes the queue and launches a go-routine to periodically invoke a
+// method to preform the rotations.
+//
+// initQueue is invoked by the Factory method in a go-routine. The Factory does
+// not wait for success or failure of it's tasks before continuing. This is to
+// avoid blocking the mount process while loading and evaluating existing roles,
+// etc.
+func (b *databaseBackend) initQueue(ctx context.Context, conf *logical.BackendConfig) {
+	// verify this mount is on the primary server, or is a local mount. If not, do
+	// not create a queue or launch a ticker. Both processing the WAL list and
+	// populating the queue are done sequentially and before launching a
+	// go-routine to run the periodic ticker.
+	replicationState := conf.System.ReplicationState()
+	if (conf.System.LocalMount() || !replicationState.HasState(consts.ReplicationPerformanceSecondary)) &&
+		!replicationState.HasState(consts.ReplicationDRSecondary) &&
+		!replicationState.HasState(consts.ReplicationPerformanceStandby) {
+		b.Logger().Info("initializing database rotation queue")
+
+		// Sleep a few seconds to allow Vault to mount and complete setup, so
+		// that we can write to storage
+		time.Sleep(3 * time.Second)
+
+		b.Lock()
+		if b.credRotationQueue == nil {
+			b.credRotationQueue = queue.New()
+		}
+		b.Unlock()
+
+		// create a context with a cancel method for processing any WAL entries and
+		// populating the queue
+		ctx, cancel := context.WithCancel(context.Background())
+		b.cancelQueue = cancel
+
+		// search through WAL for any rotations that were interrupted
+		if err := b.loadStaticWALs(ctx, conf); err != nil {
+			b.Logger().Warn("error(s) loading WAL entries into queue: ", err.Error())
+		}
+
+		// load roles and populate queue with static accounts
+		b.populateQueue(ctx, conf.StorageView)
+
+		// launch ticker
+		go b.runTicker(ctx, conf.StorageView)
+	}
+}
+
+// loadStaticWALs reads WAL entries at backend initialization. WAL entries are
+// created during the normal rotation process to ensure data integrity, in the
+// event that Vault fails to store the results (due to power failure, disk
+// failure, et. al.) If WAL entries for static account rotation are round,
+// attempt to re-set the password for the role given the NewPassword stored in
+// the WAL. If the matching Role does not exist, the Role's LastVaultRotation is
+// newer than the WAL, or the Role does not have a static account, delete the
+// WAL.
+func (b *databaseBackend) loadStaticWALs(ctx context.Context, conf *logical.BackendConfig) error {
+	keys, err := framework.ListWAL(ctx, conf.StorageView)
+	if err != nil {
+		return err
+	}
+	if len(keys) == 0 {
+		b.Logger().Info("no WAL entries found in loadStaticWALs")
+		return nil
+	}
+
+	// loop through WAL keys and process any rotation ones
+	var merr error
+	for _, walID := range keys {
+		// allow cancellation from context
+		select {
+		case <-ctx.Done():
+			b.Logger().Info("loadStaticWALs cancelled")
```

Comment:
Can we not use the function name in the error message?

Reply:
Agree, thanks. Done in 236dc1868